server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/DefaultMetaRepository.java

```diff
@@ -180,16 +183,26 @@ public DefaultMetaRepository(Repository repo) {
 
         abstract List<Mirror> toMirrors(Project parent, Iterable<MirrorCredential> credentials);
 
-        static MirrorCredential findCredential(Iterable<MirrorCredential> credentials, URI remoteUri) {
-            MirrorCredential credential = MirrorCredential.FALLBACK;
-            for (MirrorCredential c : credentials) {
-                if (c.matches(remoteUri)) {
-                    credential = c;
-                    break;
+        static MirrorCredential findCredential(Iterable<MirrorCredential> credentials, URI remoteUri,
+                                               @Nullable String credentialId) {
+            if (credentialId != null) {
+                // Find by credential ID.
+                for (MirrorCredential c : credentials) {
+                    final Optional<String> id = c.id();
+                    if (id.isPresent() && credentialId.equals(id.get())) {
```

Comment:
This is obviously a sequential search but I guess it's fine for usual use cases. Will optimize later if really necessary.